accessibility issue/2030 Removed select2 for native implementation

oliverfoster opened the pull request:
[issue/2030](https://github.com/adaptlearning/adapt_framework/issues/2030)
* Replaced select2 with native Adapt based implementation
* Select2 was horribly inaccessible and the repo has since been abandoned
* Switched to using option indexes for values instead of option texts
* Accessible, rtl

#### Testing
* Jaws + IE11 : space, up/down/home/end, space, esc (to exit forms mode)
* Jaws + Chrome: space, up/down/home/end, space

moloko commented on js/matchingView.js:
```diff
@@ -130,24 +121,27 @@ define([
 
         /**
-         * sets the selected item of a dropdown
+         * sSets the selected item of a dropdown
```
```suggestion
         * Sets the selected item of a dropdown
```

tomgreenfield commented on js/dropdownOption.js:
```diff
@@ -0,0 +1,137 @@
+        },
+
+        deselect: function() {
+            var parent = this.parent();
```
Is `parent` designed to be used?

oliverfoster replied:
i was missing a bit of code

oliverfoster commented:
All ready I believe. Thanks for reviewing.

cahirodoherty-learningpool commented on js/dropdown.js:
```diff
@@ -0,0 +1,245 @@
+                    event.preventDefault();
+                    this.$button.focus();
+                    return;
+                case 27: // ESCAPE
```
This could just be added to the cases above (13, 32)

oliverfoster replied:
Indeed it can. Ta 👍

cahirodoherty-learningpool replied:
Yeah, looks good

tomgreenfield commented on less/matching.less:
```diff
@@ -1,240 +1,111 @@
-	position: fixed;
+    .matching-select-icon {
+        &:hover {
+            color: @item-text-color-hover;
```
Ticks and crosses go white on hover.

oliverfoster replied:
this comes out of the theme now, have fixed it in the component

tomgreenfield replied:
Do we need an issue creating for this?

oliverfoster replied:
https://github.com/adaptlearning/adapt_framework/issues/2366

tomgreenfield commented on js/dropdown.js:
```diff
@@ -0,0 +1,242 @@
+                this.$list
+                    .css({
+                        top: isOffscreen ? -height : ''
+                    })
```
Suggest one-liner:
```js
.css('top', isOffScreen ? -height : '')
```

danielstorey commented on js/dropdownOption.js:
```diff
@@ -0,0 +1,141 @@
+
+        destroy: function() {
+            var parent = this.parent();
+            this.$el.remove();
```
Calling `this.remove()` accomplishes the above, but also undelegates events

oliverfoster replied:
Indeed it does. This was formerly not a view, so it's a legacy hangover from my previous implementation.

danielstorey commented on js/matchingModel.js:
```diff
@@ -65,7 +65,9 @@ define([
-            var option = item._options[optionIndex];
+            var option = _.find(item._options, function(option) {
+                return option._index === optionIndex;
+            });
```
Minor: `_.findWhere` to condense to one line

danielstorey commented on js/matchingView.js:
```diff
@@ -1,125 +1,116 @@
+            this.dropdowns.forEach(function(dropdown) {
+                if (!dropdown.isEmpty()) return;
+                dropdown.$el.parents('.matching-select-container').addClass('error');
+                return true;
```
I don't think returning a value within a `.forEach` does anything?

oliverfoster commented on js/dropdownOption.js:
```diff
@@ -0,0 +1,141 @@
+                if (option !== this) return;
+                return found = parent.options[index+1];
+            }.bind(this));
+            return found;
```
I don't think `_.find` returns value from the iterator as `_.map` does. I believe it returns the item from the list where the iterator returns a truthy. If that's the case your code will always return `parent.options[index];`

oliverfoster replied:
I've added a function called `getIndex` which simplifies it somewhat.

oliverfoster commented on js/dropdown.js:
```diff
@@ -0,0 +1,242 @@
+define([
+    './dropdownOption',
+    'handlebars'
```
Nope, it's not, I had previously had a helper definition in here.

oliverfoster commented:
@danielstorey & @tomgreenfield 
I think we're ready.

moloko commented:
In IE11, if any of the options are particularly lengthy, for example:
```json
"text": "This is a really, really stupidly long option just to test what happens when we have a really long option the quick brown fox jumps over the lazy dog"
```
then it'll be _extremely_ unreliable about letting you select that option (i.e. most of the time it doesn't let you)

moloko commented:
The current version has a fine white line at the bottom of the options 'drawer':
![matching-original](https://user-images.githubusercontent.com/1249597/52725023-5050e200-2fa8-11e9-9db2-d485799552cd.png)

this stops it from 'blending in' to subsequent options like it does now:
![matching-new](https://user-images.githubusercontent.com/1249597/52725039-5646c300-2fa8-11e9-9fd8-72dea2cc6905.png)

I think this line should be added back in

moloko commented:
Minor issue: if you have one of the dropdowns right at the bottom of the viewport, the first time you open it, it will open upwards. The second time, it will open downwards, forcing the window to scroll down.

the current version consistently opens upwards in this situation

moloko commented:
this version displays a scrollbar in the 'drawer' much more readily than the current version, for example:
![matching-scrollbar](https://user-images.githubusercontent.com/1249597/52725566-64e1aa00-2fa9-11e9-9884-55f307886710.png)

moloko commented:
on iOS (12.1.13) it's possibly to pull the entire 'drawer' down when scrolling
![matching-ios-scrolling](https://user-images.githubusercontent.com/1249597/52727524-51384280-2fad-11e9-8938-e4f356f985b4.gif)

oliverfoster commented:
> on iOS (12.1.13) it's possibly to pull the entire 'drawer' down when scrolling
> ![matching-ios-scrolling](https://user-images.githubusercontent.com/1249597/52727524-51384280-2fad-11e9-8938-e4f356f985b4.gif)

by "drawer" you mean the drop down list?

oliverfoster commented:
weirdly the gif only started playing after i quote tweeted you... so, ignore my previous comment. thanks for the gif.

moloko commented:
> weirdly the gif only started playing after i quote tweeted you

It was curiously challenging to get an animation of that happening in a format/size that github found acceptable so I'm glad it was worth it!

oliverfoster commented:
Try that @moloko 
* IE11 has slightly different focus behaviour than modern browsers, so a click on the inner span was causing the list to blur and close before the click was processed.
* iOS overscroll I didn't fix, I just added a solid background so that it's clearer that the listbox is overscrolling
* Borders I've added back in to mirror current version but I've put them in the matching code rather than the dropdown code
* List direction now properly resets so that the maths works correctly every time it's opened
* List height I've set to 50vh before it scrolls, this feels a bit more reasonable and allows a good view of the page content and list together.

tomgreenfield commented on js/dropdown.js:
```diff
@@ -0,0 +1,252 @@
+                click: this.onButtonClick
+            });
+            this.$list.on('blur', this.onListBlur);
+            this.$list.on('focusin', this.onListFocus)
```
Please switch to object style:
```js
this.$list.on({
  blur: this.onListBlur,
  focusin: this.onListFocus
});
```

tomgreenfield commented on js/dropdown.js:
```diff
@@ -0,0 +1,252 @@
+                'mousedown touchstart': this.onStartInteraction,
+                click: this.onButtonClick
+            });
+            this.$list.off('blur', this.onListBlur);
```
Also remove the `focusin` listener?

moloko commented on js/dropdown.js:
```diff
@@ -0,0 +1,251 @@
+                // the blur event will fire first closing the list.
+                // this code is to prevent the list from reopening
+                // in that situation
+                return this.wasOpen = false;
```
it's very minor but changing this to `return (this.wasOpen = false);` would stop code linters like JSHint from flagging this with a 'Did you mean to return a conditional instead of an assignment?' warning.

tomgreenfield replied:
Please put `return;` onto its own line.

oliverfoster commented:
@danielstorey could you recheck?
